codegen/method.go

```diff
@@ -679,6 +686,72 @@ func (ms *MethodSpec) setResponseHeaderFields(
 	walkFieldGroups(fields, visitor)
 }
 
+func (ms *MethodSpec) setClientRequestHeaderFields(
+	funcSpec *compile.FunctionSpec, packageHelper *PackageHelper,
+) error {
+	fields := compile.FieldGroup(funcSpec.ArgsSpec)
+
+	statements := LineBuilder{}
+	var finalError error
+	var seenOptStructs = map[string]string{}
+
+	// Scan for all annotations
+	visitor := func(
+		goPrefix string, thriftPrefix string, field *compile.FieldSpec,
+	) bool {
+		realType := compile.RootTypeSpec(field.Type)
+		longFieldName := goPrefix + "." + pascalCase(field.Name)
+
+		// If the type is a struct then we cannot really do anything
+		if _, ok := realType.(*compile.StructSpec); ok {
+			// if a field is a struct then we must do a nil check
+			typeName, err := GoType(packageHelper, realType)
+			if err != nil {
+				finalError = err
+				return true
+			}
+			seenOptStructs[longFieldName] = typeName
+			return false
+		}
+
+		if param, ok := field.Annotations[antHTTPRef]; ok {
+			if param[0:8] == "headers." {
+				headerName := param[8:]
+				bodyIdentifier := goPrefix + "." + pascalCase(field.Name)
+				var headerNameValuePair string
+				if field.Required {
+					headerNameValuePair = "headers[%q]= r%s"
+				} else {
+					headerNameValuePair = "headers[%q]= *r%s"
+				}
+				if len(seenOptStructs) == 0 {
+					statements.appendf(headerNameValuePair,
+						headerName, bodyIdentifier,
+					)
+				} else {
+					for seenStruct := range seenOptStructs {
+						if strings.HasPrefix(longFieldName, seenStruct) {
+							statements.appendf("if r%s != nil {", seenStruct)
+							statements.appendf(headerNameValuePair,
```

Comment:
I dont think this will work if `len(seenOptStruct) >= 2`

That would generate two if statements and two assignment expressions.

We want to have two loops I guess; one to open the if statements and one to close them.